src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+            if (!dry_run)
+                data = jitc_malloc(alloc_type, dsize);
+        } else if (this->alloc_size < dsize) {
+            this->alloc_size = dsize;
+            jitc_log(LogLevel::Debug, "    re-allocating output of size %zu.",
+                     dsize);
+            if (!dry_run) {
+                jitc_free(this->data);
+                data = jitc_malloc(alloc_type, dsize);
+            }
+        } else {
+            // Do not reallocate if the size is enough
+        }
+
+        this->data_size = dsize;
+    }
+
+    void free() {
+        jitc_free(this->data);
+        this->data       = nullptr;
+        this->data_size  = 0;
+        this->alloc_size = 0;
+    }
+};
+
+/// Kernel parameter buffer
+static std::vector<void *> kernel_params;
+
+/// Temporary variables used for replaying a recording.
+static std::vector<ReplayVariable> replay_variables;
+
+static ProfilerRegion pr_operation("Replay Operation");
+static ProfilerRegion pr_kernel_launch("KernelLaunch");
+static ProfilerRegion pr_barrier("Barrier");
+static ProfilerRegion pr_memset_async("MemsetAsync");
+static ProfilerRegion pr_reduce_expanded("ReduceExpanded");
+static ProfilerRegion pr_expand("Expand");
+static ProfilerRegion pr_compress("Compress");
+static ProfilerRegion pr_memcpy_async("MemcpyAsync");
+static ProfilerRegion pr_mkperm("Mkperm");
+static ProfilerRegion pr_block_reduce("BlockReduce");
+static ProfilerRegion pr_block_prefix_reduce("BlockPrefixReduce");
+static ProfilerRegion pr_aggregate("Aggregate");
+static ProfilerRegion pr_free("Free");
+static ProfilerRegion pr_output("Output");
+
+int Recording::replay(const uint32_t *replay_inputs, uint32_t *outputs) {
+
+    uint32_t n_kernels = 0;
+
+    ThreadState *ts = thread_state(backend);
+    if (dynamic_cast<RecordThreadState *>(ts) != nullptr)
+        jitc_raise("replay(): Tried to replay while recording!");
+
+    OptixShaderBindingTable *tmp_sbt = ts->optix_sbt;
+    scoped_set_context_maybe guard2(ts->context);
+
+    replay_variables.clear();
+
+    replay_variables.reserve(this->record_variables.size());
+    for (RecordVariable &rv : this->record_variables) {
+        replay_variables.push_back(ReplayVariable(rv));
+    }
+
+    jitc_log(LogLevel::Info, "replay(): inputs");
+    // Populate with input variables
+    for (uint32_t i = 0; i < this->inputs.size(); ++i) {
+        Variable *input_variable = jitc_var(replay_inputs[i]);
+        replay_variables[this->inputs[i]].init_from_input(input_variable);
+        jitc_log(LogLevel::Debug, "    input %u: r%u maps to slot s%u", i,
+                 replay_inputs[i], this->inputs[i]);
+    }
+
+    // The main loop that executes each operation
+    for (uint32_t i = 0; i < this->operations.size(); ++i) {
+        Operation &op = this->operations[i];
+        ProfilerPhase profiler(pr_operation);
+        if (!op.enabled)
+            continue;
+
+        switch (op.type) {
+            case OpType::KernelLaunch: {
```

Comment:
Move this case's handling to a function, since it's quite involved?
The same is done in e.g. `cuda_eval.cpp`, there are local functions for the more complex cases.

Reply:
They are now ordered in the following way (mkperm example):
```cpp
uint32_t RecordThreadState::mkperm(...){...}
void RecordThreadState::record_mkperm(...){...}
void RecordThreadState::replay_mkperm(...){...}
```